File: gatling-http/src/main/scala/io/gatling/http/request/builder/ws/Ws.scala
/**
 * Copyright 2011-2016 GatlingCorp (http://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gatling.http.request.builder.ws

import io.gatling.core.session.{ Expression, SessionPrivateAttributes }
import io.gatling.http.action.async.ws._
import io.gatling.http.check.async.AsyncCheckBuilder
import io.gatling.http.request.builder.CommonAttributes

object Ws {

  val DefaultWebSocketName = SessionPrivateAttributes.PrivateAttributePrefix + "http.webSocket"
}

/**
 * @param requestName The name of this request
 * @param wsName The name of the session attribute used to store the websocket
 */
class Ws(requestName: Expression[String], wsName: String = Ws.DefaultWebSocketName) {

  def wsName(wsName: String) = new Ws(requestName, wsName)

  /**
   * Opens a web socket and stores it in the session.
   *
   * @param url The socket URL
   *
   */
  def open(url: Expression[String]) = new WsOpenRequestBuilder(CommonAttributes(requestName, "GET", Left(url)), wsName)

  /**
   * Sends a binary message on the given websocket.
   *
   * @param bytes The message
   */
  def sendBytes(bytes: Expression[Array[Byte]]) = new WsSendBuilder(requestName, wsName, bytes.map(BinaryMessage))

  /**
   * Sends a text message on the given websocket.
   *
   * @param text The message
   */
  def sendText(text: Expression[String]) = new WsSendBuilder(requestName, wsName, text.map(TextMessage))

  /**
   * Check for incoming messages on the given websocket.
   *
   * @param checkBuilder The check builder
   */
  def check(checkBuilder: AsyncCheckBuilder) = new WsSetCheckBuilder(requestName, checkBuilder, wsName)

  /**
   * Cancel current check on the given websocket.
   *
   */
  def cancelCheck = new WsCancelCheckBuilder(requestName, wsName)

  /**
   * Reconciliate the main state with the one of the websocket flow.
   */
  def reconciliate = new WsReconciliateBuilder(requestName, wsName)

  /**
   * Closes a websocket.
   */
  def close = new WsCloseBuilder(requestName, wsName)
}

File: gatling-http-java/src/main/java/io/gatling/javaapi/http/Ws.java
/*
 * Copyright 2011-2024 GatlingCorp (https://gatling.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.gatling.javaapi.http;

import static io.gatling.javaapi.core.internal.Expressions.*;

import edu.umd.cs.findbugs.annotations.NonNull;
import io.gatling.javaapi.core.ActionBuilder;
import io.gatling.javaapi.core.Session;
import java.util.function.Function;

/**
 * DSL for building WebSocket configurations
 *
 * <p>Immutable, so all methods return a new occurrence and leave the original unmodified.
 */
public final class Ws {
  private final io.gatling.http.action.ws.Ws wrapped;

  Ws(final io.gatling.http.action.ws.Ws wrapped) {
    this.wrapped = wrapped;
  }

  /**
   * Define a custom WebSocket name so multiple WebSockets for the same virtual users don't conflict
   *
   * @param wsName the name, expressed as a Gatling Expression Language String
   * @return a new Ws instance
   */
  @NonNull
  public Ws wsName(@NonNull String wsName) {
    return new Ws(wrapped.wsName(toStringExpression(wsName)));
  }

  /**
   * Define a custom WebSocket name so multiple WebSockets for the same virtual users don't conflict
   *
   * @param wsName the name, expressed as a function
   * @return a new Ws instance
   */
  @NonNull
  public Ws wsName(@NonNull Function<Session, String> wsName) {
    return new Ws(wrapped.wsName(javaFunctionToExpression(wsName)));
  }

  /**
   * Boostrap an action to connect the WebSocket
   *
   * @param url the url to connect to, expressed as a Gatling Expression Language String
   * @return the next DSL step
   */
  @NonNull
  public WsConnectActionBuilder connect(@NonNull String url) {
    return new WsConnectActionBuilder(wrapped.connect(toStringExpression(url)));
  }

  /**
   * Boostrap an action to connect the WebSocket
   *
   * @param url the url to connect to, expressed as a function
   * @return the next DSL step
   */
  @NonNull
  public WsConnectActionBuilder connect(@NonNull Function<Session, String> url) {
    return new WsConnectActionBuilder(wrapped.connect(javaFunctionToExpression(url)));
  }

  /**
   * Boostrap an action to send a TEXT frame
   *
   * @param text the text to send, expressed as a Gatling Expression Language String
   * @return the next DSL step
   */
  @NonNull
  public WsSendTextActionBuilder sendText(@NonNull String text) {
    return new WsSendTextActionBuilder(wrapped.sendText(toStringExpression(text)));
  }

  /**
   * Boostrap an action to send a TEXT frame
   *
   * @param text the text to send, expressed as a function
   * @return the next DSL step
   */
  @NonNull
  public WsSendTextActionBuilder sendText(@NonNull Function<Session, String> text) {
    return new WsSendTextActionBuilder(wrapped.sendText(javaFunctionToExpression(text)));
  }

  /**
   * Boostrap an action to send a BINARY frame
   *
   * @param bytes the static bytes to send
   * @return the next DSL step
   */
  @NonNull
  public WsSendBinaryActionBuilder sendBytes(@NonNull byte[] bytes) {
    return new WsSendBinaryActionBuilder(wrapped.sendBytes(toStaticValueExpression(bytes)));
  }

  /**
   * Boostrap an action to send a BINARY frame
   *
   * @param bytes the bytes to send, expressed as a Gatling Expression Language String
   * @return the next DSL step
   */
  @NonNull
  public WsSendBinaryActionBuilder sendBytes(@NonNull String bytes) {
    return new WsSendBinaryActionBuilder(wrapped.sendBytes(toBytesExpression(bytes)));
  }

  /**
   * Boostrap an action to send a BINARY frame
   *
   * @param bytes the bytes to send, expressed as a function
   * @return the next DSL step
   */
  @NonNull
  public WsSendBinaryActionBuilder sendBytes(@NonNull Function<Session, byte[]> bytes) {
    return new WsSendBinaryActionBuilder(wrapped.sendBytes(javaFunctionToExpression(bytes)));
  }

  /**
   * Boostrap an action to send a CLOSE frame with the default 1000 status code
   *
   * @return the next DSL step
   */
  @NonNull
  public ActionBuilder close() {
    return wrapped::close;
  }

  /**
   * Boostrap an action to send a CLOSE frame with specified status and reason
   *
   * @param statusCode the close frame status code
   * @param reason the close frame reason
   * @return the next DSL step
   */
  @NonNull
  public ActionBuilder close(int statusCode, String reason) {
    return () -> wrapped.close(statusCode, reason);
  }

  public static final class Prefix {

    public static final Prefix INSTANCE = new Prefix();

    private Prefix() {}

    /**
     * Bootstrap a check on inbound TEXT frames
     *
     * @param name the name of the check, expressed as a Gatling Expression Language String
     * @return the next DSL step
     */
    @NonNull
    public WsFrameCheck.Text checkTextMessage(@NonNull String name) {
      return new WsFrameCheck.Text(
          io.gatling.http.Predef.ws().checkTextMessage(toStringExpression(name)));
    }

    /**
     * Bootstrap a check on inbound TEXT frames
     *
     * @param name the name of the check, expressed as a function
     * @return the next DSL step
     */
    @NonNull
    public WsFrameCheck.Text checkTextMessage(@NonNull Function<Session, String> name) {
      return new WsFrameCheck.Text(
          io.gatling.http.Predef.ws().checkTextMessage(javaFunctionToExpression(name)));
    }

    /**
     * Bootstrap a check on inbound BINARY frames
     *
     * @param name the name of the check, expressed as a Gatling Expression Language String
     * @return the next DSL step
     */
    @NonNull
    public WsFrameCheck.Binary checkBinaryMessage(@NonNull String name) {
      return new WsFrameCheck.Binary(
          io.gatling.http.Predef.ws().checkBinaryMessage(toStringExpression(name)));
    }

    /**
     * Bootstrap a check on inbound BINARY frames
     *
     * @param name the name of the check, expressed as a function
     * @return the next DSL step
     */
    @NonNull
    public WsFrameCheck.Binary checkBinaryMessage(@NonNull Function<Session, String> name) {
      return new WsFrameCheck.Binary(
          io.gatling.http.Predef.ws().checkBinaryMessage(javaFunctionToExpression(name)));
    }
  }
}
